### DISCUSSION TITLE
SSL extension that does TCP MLLP

### DISCUSSION THREAD

**Discussion by Wet-Bandit** [nextgenhealthcare/connect#4885] [Q&A]
Has anyone out there wrote an SSL mirth connect extension, that's available and does the TCP_IP source connector?
Or has anyone out there already updated this mirth connect extension below to do TCP MLLP in addition to HTTPS?
https://github.com/tonygermano/connect-plugins

**Comment by marcheschi** [comment]
Hi
You can use [stunnel](https://www.stunnel.org) to implement mllp over ssl.
Paolo

**Reply by Wet-Bandit** [reply]
The customer Tech department no longer allows the use of Stunnel.  The application or (ESB) must implement SSL itself.

**Reply by tonygermano** [reply]
Sounds like you need to suck it up and buy one of the commercial plugins. They aren't that expensive. The nextgen one is the least expensive if you only need to license one mirth server, and it comes with the channel history plugin as well. The Zen SSL plugin gets more cost effective when you have multiple production servers to license.

**Reply by pacmano1** [reply]
Also - if you expect all your partner organizations with whom you integrate to natively wrap TCP in TLS, you are in for a very rude surprise.

**Reply by Wet-Bandit** [reply]
Not sure why you are talking to me rudely.  I don't have the ability to purchase plug-ins.  I have made mentioned that before.  Nor do I have any power to make the customer make purchases.  And, I am obligated to follow the requirements of my customer.  And they have clearly banned the use of Stunnel for all servers as our servers are being moved to a cloud service.  Your comment is not going to help.  I personally have no problem purchasing an SSL plug-in.  But I am not going to fork out $2-$4k of my own personal money every year for my customer, nor am I allowed to.  As for the TCP in TLS, that is an internal connection that the customer has (that I never agreed with such design in the first place).  They didn't want mirth connect channels exposed to the internet so we have a java https web service that forwards to mirth channels via MLLP/TCP.  I appreciate your enthusiasm, but I need to do what the customer asks.

**Reply by Wet-Bandit** [reply]
FYI:  I have already told the customer my suggestion and opinion was to purchase a yearly license of the mirth-nextgen SSL plug-in at least 3 times already.  One reply was, that it takes too long for a purchase like that to go through the system (could be 3-6 months) and we need our servers on the cloud by the end of the year.

**Reply by pacmano1** [reply]
No one is talking to you rudely.   People are giving you advice that you can take or leave.   You requested a free solution and folks proposed haproxy and stunnel and your response was "can't use that".  Then a paid plugin was suggested and you said "can't do that either".

So that suggests the following:

1) There are other solutions to encrypt that traffic - implement IPSec between your servers.  
2) Moving to AWS does not fix your problem if you continue to use Mirth as you know.
3) If you have "... https web service that forwards to mirth channels via MLLP/TCP",  seems like you might have development talent to write your own plugin and share that back to the community - since you and the state are paying nothing for mirth or so it seems.

Frankly, your tone here and on the forums isn't something that motivates me anymore to provide any help on any of your posts anymore.  

Best wishes with your search for this solution and others, I'm out.

**Reply by Wet-Bandit** [reply]
Please only post if you have a working suggestion to the posted problem and scenario.  Rude comments and then being defensive because you were called out on being rude is not allowed in the forum.

**Reply by Wet-Bandit** [reply]
Problem was resolved last month.  I just did the following:  Remove MLLP characters (via simple .trim()) and add a simple HTTP header to the HL7 message and use a secure socket class to send the message to the mirth connect port.  Then change the receiving mirth channel source connector from TCP_IP MLLP to HTTP.  Using the tonygermano https://github.com/tonygermano/connect-plugins      the connection will be HTTPS with SSL, then remove the HTTP header on the response and send back the clean HL7.  Hours of research, 10 minutes of coding.  All the more reason SSL should just be included in mirth connect.  Also you will have to update the plug-in (written in python) and re-compile, it was missing a method signature or something.  Someone else in another group did this and then forwarded the plug-in to us.

**Reply by pacmano1** [reply]
For others reading this thread  - the solution posted here does not address the encryption of TCP MLLP connections which is a transport level problem.   Instead the solution wraps the HL7 dialog on a different transport (https).    Nothing wrong with that of course, it just doesn't address the OPs original request to do "**TCP MLLP** _in addition to_ HTTPS"

**Reply by Wet-Bandit** [reply]
The code to do this is super simple.  Read below and assume you have a String object called message that contains the HL7...
(Java code)
//Add HTTP header
String httpMessage="POST / HTTP/1.0\r\nContent-Length: "+message.length()+"\r\nContent-Type: x-application/hl7-v2+er7\r\n\r\n"+message;
//Remove HTTP header
(strResponse.substring(strResponse.indexOf("\r\n\r\n")+4));

**Reply by Wet-Bandit** [reply]
RE:  "If you have "... https web service that forwards to mirth channels via MLLP/TCP", seems like you might have development talent to write your own plugin and share that back to the community - since the customer is paying nothing for mirth connect." 
 
Yes, I probably could update the plug-in or write a new one that does SSL in MLLP/TCP_IP since I write the web service etc.  and I definitely  would provide it to the community for free.  But since my customer's request is resolved I have to work on what they want done next.  

It also seems frivolous to do such work when the tool should just provide this modern commonplace ability to use SSL etc..  Nowadays everything is SSL and secure as the "norm".  How about someone gives you a free car that came with the condition that you had to add your own brakes?

**Reply by pacmano1** [reply]
That plugin is a revenue stream for Nextgen.   It's their business - people don't have to use their product at all if they don't like that that SSL isn't bundled.   The core extension pack also has the channel history plugin, which is a life saver.

And there are workarounds - the most obvious one supported by all operating system is IPSec (and in the work our company does is the norm for securing HL7v2 over TCP) - which encrypts all traffic between nodes as are the other solutions you used and others mentioned.   You can of course use none of the solutions here - and instead encrypt the _contents_ of the message (still wrapping it in MLLP) and decrypting the content on the remote side - that meets the intention of HIPAA, which is vague in defining the technology required to secure PII/PHI.

**Reply by Wet-Bandit** [reply]
> Sounds like you need to suck it up and buy one of the commercial plugins. They aren't that expensive. The nextgen one is the least expensive if you only need to license one mirth server, and it comes with the channel history plugin as well. The Zen SSL plugin gets more cost effective when you have multiple production servers to license.

tonygermano:
Hello.  You are the one who wrote the free SSL extension for mirth connect.  Thank you on behalf of myself and a few others.  You saved our skin, our projects, and possibly even our (contracts/jobs/customers).  Hope you will be blessed in some fashion.  Besides updating the methods signatures re-compiling the Python code, it seems to be working fine as is.  No need to buy anything.  Once again, much thanks for providing what mirth connect did not.

**Reply by Wet-Bandit** [reply]
Nextgen is the revenue stream, that's why they bought Mirth Corporation to save all of it's hard work and own a great "Health Industry" ESB tool.  The plug-in extension is not a revenue stream when no one buys it.  Rather it appears to be hindering the expanding usage and growth of mirth connect.
